### PR TITLE
[#166739839] User Metadata API

### PR DESCRIPTION
**Pull request by BurnedMarshal**

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Please include a Pivotal story at the beginning of the PR title (see below).
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  Please include a description of your PR changes.
  </td>
    </tr>
  </tbody>
</table>



  Example of PR titles that include pivotal stories:

  * single story: `[#123456] my PR title`
  * multiple stories: `[#123456,#123457,#123458] my PR title`

    
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: src/services/redisUserMetadataStorage.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    ) {
+      return left(invalidVersionNumberError);
+    }
+    if (
+      isLeft(getUserMetadataResult) &&
+      getUserMetadataResult.value !== metadataNotFoundError
+    ) {
+      return left(getUserMetadataResult.value);
+    }
+    return await new Promise<Either<Error, boolean>>(resolve => {
```
Comment (cloudify) [inline src/services/redisUserMetadataStorage.ts]: this is not the correct semantic, if two clients attempt to set a certain version concurrently, they will both succeed

you need to implement [compare-and-set](https://redis.io/topics/transactions#cas)

File: src/services/redisUserMetadataStorage.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import { UserMetadata } from "../../generated/backend/UserMetadata";
+import { log } from "../utils/logger";
+import { IUserMetadataStorage } from "./IUserMetadataStorage";
+import RedisStorageUtils from "./redisStorageUtils";
+
+const userMetadataPrefix = "USERMETA-";
+export const metadataNotFoundError = new Error("User Metadata not found");
+export const invalidVersionNumberError = new Error("Invalid version number");
+
+export default class RedisUserMetadataStorage extends RedisStorageUtils
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/redisUserMetadataStorage.ts]: fixed in 6428fbf

File: src/services/redisStorageUtils.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Parse the a Redis integer reply.
+   *
+   * @see https://redis.io/topics/protocol#integer-reply
+   */
+  protected integerReply(
+    err: Error | null,
+    // tslint:disable-next-line:no-any
+    reply: any
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/redisStorageUtils.ts]: fixed in 6428fbf

File: src/services/redisStorageUtils.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  ): Either<Error, boolean> {
+    if (err) {
+      return left<Error, boolean>(err);
+    }
+
+    return right<Error, boolean>(reply === "OK");
+  }
+
+  /**
+   * Parse the a Redis integer reply.
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/redisStorageUtils.ts]: fixed in 6428fbf

File: src/services/redisStorageUtils.ts [hunk 5]
```diff
@@ -0,0 +1,40 @@
+import { Either, left, right } from "fp-ts/lib/Either";
+import { isNumber } from "util";
+
+export default class RedisStorageUtils {
+  // tslint:disable-next-line: no-empty
+  constructor() {}
+
+  /**
+   * Parse the a Redis single string reply.
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/redisStorageUtils.ts]: fixed in 6428fbf

File: src/services/redisStorageUtils.ts [hunk 6]
```diff
@@ -0,0 +1,40 @@
+import { Either, left, right } from "fp-ts/lib/Either";
+import { isNumber } from "util";
+
+export default class RedisStorageUtils {
+  // tslint:disable-next-line: no-empty
+  constructor() {}
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/redisStorageUtils.ts]: fixed in 6428fbf

File: src/services/IUserMetadataStorage.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+ * Interface for the session storage services.
+ */
+
+import { Either } from "fp-ts/lib/Either";
+import { UserMetadata } from "generated/backend/UserMetadata";
+import { User } from "../types/user";
+
+export interface IUserMetadataStorage {
+  /**
+   * Stores a User Metadata information from Redis Storage
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/IUserMetadataStorage.ts]: fixed in 6428fbf

File: src/services/IUserMetadataStorage.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+   */
+  readonly set: (
+    user: User,
+    payload: UserMetadata
+  ) => Promise<Either<Error, boolean>>;
+
+  /**
+   * Retreive User Metadata information from Redis Storage
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/services/IUserMetadataStorage.ts]: fixed in 6428fbf

File: src/controllers/userMetadataController.ts [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  ): Promise<
+    | IResponseErrorValidation
+    | IResponseErrorInternal
+    | IResponseSuccessJson<UserMetadataResponse>
+  > =>
+    withUserFromRequest(req, async user => {
+      const metadata = await this.userMetadataStorage.get(user);
+      if (isLeft(metadata)) {
+        return metadata.value === metadataNotFoundError
+          ? ResponseSuccessJson({})
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/controllers/userMetadataController.ts]: fixed in 6428fbf

File: api_backend.yaml [hunk 10]
```diff
@@ -638,6 +680,22 @@ definitions:
...
+    title: User Metadata information
+    properties:
+      version:
+        type: number
+      metadata:
+        type: string
+    required:
+      - version
+      - metadata
+  UserMetadataResponse:
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply api_backend.yaml]: fixed in 6428fbf

File: api_backend.yaml [hunk 11]
```diff
@@ -293,6 +293,48 @@ paths:
...
+    x-swagger-router-controller: userMetadataController
+    get:
+      operationId: getUserMetadata
+      summary: Get user's metadata
+      description: Returns metadata for the user identified by the provided fiscal code.
+      responses:
+        "200":
+          description: Found.
+          schema:
+            $ref: "#/definitions/UserMetadataResponse"
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply api_backend.yaml]: fixed in 6428fbf

File: src/controllers/userMetadataController.ts [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+  metadataNotFoundError
+} from "../services/redisUserMetadataStorage";
+import { withUserFromRequest } from "../types/user";
+import { withValidatedOrValidationError } from "../utils/responses";
+
+export default class UserMetadataController {
+  constructor(private readonly userMetadataStorage: IUserMetadataStorage) {}
+
+  /**
+   * Returns the metadata for the user identified by the provided fiscal
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/controllers/userMetadataController.ts]: fixed in 6428fbf

File: api_backend.yaml [hunk 13]
```diff
@@ -293,6 +293,48 @@ paths:
...
+        "401":
+          description: Bearer token null or expired.
+        "500":
+          description: There was an error in retrieving the user metadata.
+          schema:
+            $ref: "#/definitions/ProblemJson"
+    post:
+      operationId: upsertUserMetadata
+      summary: Set User's metadata
+      description: Create or update metadata for the user identified by the provided fiscal code.
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply api_backend.yaml]: fixed in 6428fbf

File: api_backend.yaml [hunk 14]
```diff
@@ -293,6 +293,48 @@ paths:
           description: Profile cannot be updated.
           schema:
             $ref: "#/definitions/ProblemJson"
+  "/user-metadata":
+    x-swagger-router-controller: userMetadataController
+    get:
+      operationId: getUserMetadata
+      summary: Get user's metadata
+      description: Returns metadata for the user identified by the provided fiscal code.
```
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply api_backend.yaml]: fixed in 6428fbf

File: src/services/redisUserMetadataStorage.ts (hunk 1 above)
Comment (BurnedMarshal) [reply src/services/redisUserMetadataStorage.ts]: I understand, but a solution is not so simple. The value stored on redis is not a simple integer value, but a stringified JSON having the property `version: number`. Nevertheless the specific API call works on a resource shared only with devices of the same user, concurrency on storage with different metadata values is a rare scenario.

File: src/controllers/userMetadataController.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (BurnedMarshal) [reply src/controllers/userMetadataController.ts]: `italia-ts-commons` library doesn't include `NoRespose` method, i've implemented it into `utils/responses` of this project. Is it correct?

File: src/services/redisUserMetadataStorage.ts (hunk 1 above)
Comment (cloudify) [reply src/services/redisUserMetadataStorage.ts]: nobody said it was simple :D

ok so add a BIG comment explaining that the operation is not safe for concurrent updates, also create a story in the backend backlog for implementing proper CAS and link to that story from the comment here

Comment (BurnedMarshal) [reply src/services/redisUserMetadataStorage.ts]: Fix will be implemented in [this](https://www.pivotaltracker.com/story/show/167064659) story

Comment (BurnedMarshal) [reply src/services/redisUserMetadataStorage.ts]: Code comments added in commit fc65a2a

**Comment by cloudify** [comment]
@fpersico please check if this serves your needs

**Comment by cloudify** [comment]
@BurnedMarshal could you please rebase this PR? thx

**Pull request merged**